hangouts_chat.py

```diff
@@ -13,47 +14,124 @@
 log = logging.getLogger('errbot.backends.hangoutschat')
 
 
-def _get_authenticated_http_client(creds_file, scope='https://www.googleapis.com/auth/chat.bot'):
-    return _get_google_credentials(creds_file, scope).authorize(httplib2.Http())
-
-
-def _get_google_credentials(creds_file, scope):
-    return ServiceAccountCredentials.from_json_keyfile_name(creds_file, scopes=[scope])
-
-
 class RoomsNotSupportedError(RoomError):
     def __init__(self, message=None):
         if message is None:
             message = (
-                "Room Operations are not supported in Google Hangouts Chat."
+                "Most Room operations are not supported in Google Hangouts Chat."
                 "While Rooms are a _concept_, the API is minimal and does not "
                 "expose this functionality to bots"
             )
         super().__init__(message)
 
 
+class GoogleHangoutsChatAPI:
+    """
+    Represents the Google Hangouts REST API
+    See: https://developers.google.com/hangouts/chat/reference/rest/
+    """
+    base_url = 'https://chat.googleapis.com/v1'
+    # Numbe of results to fetch at a time. Default is 100, Max is 1000
+    page_size = 500
+
+    def __init__(self, creds_file: str, scope: str = 'https://www.googleapis.com/auth/chat.bot'):
+        self.creds_file = creds_file
+        self.scope = scope
+
+    @property
+    def credentials(self):
+        return ServiceAccountCredentials.from_json_keyfile_name(self.creds_file,
+                                                                scopes=[self.scope])
+
+    @property
+    def client(self):
+        return self.credentials.authorize(httplib2.Http())
+
+    def _get(self, uri: str, query_string: str = None) -> Optional[dict]:
+        url = '{}/{}'.format(self.base_url, uri)
+        if query_string:
+            url += '?{}'.format(query_string)
+        result, content = self.client.request(uri=url, method='GET')
+        if result['status'] == "200":
+            content_json = json.loads(content.decode('utf-8'))
+            return content_json
+        else:
+            log.error('status: {}, content: {}'.format(result['status'], content))
+
+    def _post(self, uri: str, body: dict, query_string: str = None) -> Optional[dict]:
+        url = '{}/{}'.format(self.base_url, uri)
+        if query_string:
+            url += '?{}'.format(query_string)
+        result, content = self.client.request(
+            uri=url,
+            method='POST',
+            headers={'Content-Type': 'application/json; charset=UTF-8'},
+            body=json.dumps(body))
+        if result['status'] == "200":
```

Comment:
This logic feels repeated. Could we split it out to a generic `_request` method?

Reply:
Done :)